### ISSUE TITLE
Drop support for obsolete algorithms and protocols

### ISSUE THREAD

**Issue by DemiMarie** [openssl/openssl#14205]
Several algorithms and protocols supported by OpenSSL are obsolete.  These include (but are not limited to)

- RC5
- RC4
- MD2
- MD4 (sadly still used by NTLM compat stuff)
- RIPE-MD
- SSL (all versions)
- TLSv1.0
- TLSv1.1
- Blowfish (inherently vulnerable to timing attacks)
- IDEA

Other algorithms that might be removed:

- Camellia
- TDES
- Probably others I cannot think of

**Comment by mattcaswell** [comment]
> RC5
RC4
MD2
MD4 (sadly still used by NTLM compat stuff)
RIPE-MD
Blowfish (inherently vulnerable to timing attacks)
IDEA

There are currently two ways to access these algorithms:

1) Via low-level APIs
2) Via the high-level "EVP" APIs

In the forthcoming OpenSSL 3.0 all the low-level APIs for these algorithms have been deprecated and will be removed entirely from some future version of OpenSSL in accordance with our stability policy.

For access via the high-level "EVP" APIs, all of these algorithms have been moved to our "legacy" provider in OpenSSL 3.0 meaning that they are not loaded or available by default at runtime through this route.

> SSL (all versions)

SSLv2 was removed from 1.1.0. SSLv3 is still available in the codebase but is not compiled in by default. You have to explicitly request that support is compiled in if you wish to use it.

> TLSv1.0
TLSv1.1

These are not yet scheduled for removal. We continue to review these things each time we do a new release.

**Comment by DemiMarie** [comment]
It would be nice if RSA key encipherment could be turned off in `openssl.cnf`.

**Comment by kaduk** [comment]
> > TLSv1.0
TLSv1.1

> These are not yet scheduled for removal. We continue to review these things each time we do a new release.

Note that draft-ietf-tls-oldversions-deprecate is in the RFC Editor's queue and should be published in the next month or so.

**Comment by richsalz** [comment]
And the IETF isn't the protocol police :)

**Comment by DemiMarie** [comment]
> And the IETF isn't the protocol police :)

[To quote Adam Langley](https://www.imperialviolet.org/2014/12/08/poodleagain.html):

> This seems like a good moment to reiterate that _everything_ less than TLS 1.2 with an AEAD cipher suite is cryptographically broken.

Having support for broken protocols in OpenSSL does not appear to be a good idea.  At the very least, they should be disabled by default, but I would prefer them to be removed entirely.

**Comment by jyxjjj** [comment]
But I think we may just disable it by default instead of delete.
There are some of my users like keep openssl up-to-date but using these old and unsafety algo.
Otherwise a lot of developers like me must alert our users.
And, what will happen to the websites like SSLLabs?

**Comment by DemiMarie** [comment]
> But I think we may just disable it by default instead of delete. There are some of my users like keep openssl up-to-date but using these old and unsafety algo. Otherwise a lot of developers like me must alert our users. And, what will happen to the websites like SSLLabs?

Why are they using those algorithms?  And SSLLabs can use a bespoke implementation in a memory safe language.

**Comment by thejsa** [comment]
> > But I think we may just disable it by default instead of delete. There are some of my users like keep openssl up-to-date but using these old and unsafety algo. Otherwise a lot of developers like me must alert our users. And, what will happen to the websites like SSLLabs?
> 
> Why are they using those algorithms? And SSLLabs can use a bespoke implementation in a memory safe language.

Hope it’s alright for me to chime in here — I still need to use SSLv3 and RC4 in order to host homebrew reimplementations of online play services for Nintendo DS consoles.

For this use case transport security isn’t a concern (especially since there are separate vulnerabilities in cert verification), but there’s no chance of updating client software shipped on game cartridges in 2004 unfortunately.

I can understand if niche use cases like this make it not worth the effort to maintain, and building a leaner, bespoke implementation on our end is desirable; we just don’t have the resources as a small hobby project to do so at the moment, so would have to rely on older versions of OpenSSL should these protocols be dropped entirely.

**Comment by sudofox** [comment]
I'm also maintaining a service for Nintendo DS for which I cannot update the client SSL stack, similar to the above comment. SSLv3 and RC4-SHA, which are enabled on a very strict basis for where they're being used. Not being able to use these would force us to run outdated software.

**Comment by jyxjjj** [comment]
> @jyxjjj do you have Ed25519 or ECDSA host keys on the server?

I have but i commented it via the sshd_config

**Comment by DemiMarie** [comment]
> > @jyxjjj do you have Ed25519 or ECDSA host keys on the server?
> 
> I have but i commented it via the sshd_config

Turning those back on should fix the problem.

**Comment by jyxjjj** [comment]
No, it doesn't work. I've tried. If i remove the ssh-rsa from PubkeyAcceptedAlgorithms, whatever i set Ed25519 or ECDSA HostKeys commented or uncommented, it doesn't work.

**Comment by DemiMarie** [comment]
> No, it doesn't work. I've tried. If i remove the ssh-rsa from PubkeyAcceptedAlgorithms, whatever i set Ed25519 or ECDSA HostKeys commented or uncommented, it doesn't work.

What do you get in the OpenSSH logs?

**Comment by jyxjjj** [comment]
> > No, it doesn't work. I've tried. If i remove the ssh-rsa from PubkeyAcceptedAlgorithms, whatever i set Ed25519 or ECDSA HostKeys commented or uncommented, it doesn't work.
> 
> What do you get in the OpenSSH logs?

Client is JetBrains PhpStorm SFTP Connector with SSHJ, it reports 
```
Unable to reach a settlement: [ssh-ed25519, ecdsa-sha2-nistp256, ecdsa-sha2-nistp384, ecdsa-sha2-nistp521, ssh-rsa, ssh-rsa-cert-v01@openssh.com, ssh-dss] and [rsa-sha2-512, rsa-sha2-256]
```
And the server's, i don't know where to find.

**Comment by jyxjjj** [comment]
I tried again, and if i disable RSA host key, it works.
As the JetBrains YouTrack said, this may be an issue that SSHJ's.

**Comment by nhorman** [comment]
Marking as inactive, to be closed at the end of 3.4 dev barring further input

**Comment by richsalz** [comment]
The original plan was to move obsolete algorithms to a separate provider.  Has that plan been dropped?
Please don't close this but move it to community or such.

**Comment by nhorman** [comment]
No, its something we would like to do,  but the malleable definition "obsolete" along with the unknown fog of which users use said algorithms beyond that definition make their migration difficult from a support and user experience standpoint.

We've begun work on that here: https://github.com/openssl/general-policies/pull/67
by creating a policy and mechanism to notify users of how we plan to do that deprecation and movement.

Once that is approved, we have the confidence to move forward by giving users a fair chance to recognize the change and adjust their provider configuration accordingly.

Suggest that, after its approval, separate issues be opened for specific algorithms that should be considered for migration, so they can be independently considered, and addressed via the proposed process.

**Comment by richsalz** [comment]
Thanks for the info.  I think you should keep this open until the policy/mechanisms are defined and then close this with a note suggesting creating specific issues.  Keep this open since the project *is* working on it.

**Comment by nhorman** [comment]
okie dokey, though I would prefer to see separate issues for each alg that we want to deprecate, solely too be able to separately address them

**Comment by richsalz** [comment]
Separate issues makes sense.  But until you have the policy you can't request that people do that :)

**Comment by nhorman** [comment]
Agreed, thats why I'm leaving this issue in its current state until we get there, to be revisited when the policy is in place and we do have those issues opened.

**Comment by t8m** [comment]
This is too broad and useless issue. Any deprecation/move to legacy/removal of an algorithm should be discussed separately.
